### PR TITLE
Add storage point tracking and tie in storage rewards to economics

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem

Storage point values were based on a placeholder value

#### Summary of Changes

- Tie in storage rewards to economics
- Track rewards per epoch
- Allow reward collection after an epoch

### CODE REVIEW

File: programs/storage_api/src/storage_contract.rs [hunk 1]
```diff
@@ -359,13 +376,15 @@ impl<'a> StorageAccount<'a> {
     pub fn claim_storage_reward(
         &mut self,
         rewards_pool: &mut KeyedAccount,
+        current: Current,
```
Comment (rob-solana) [inline programs/storage_api/src/storage_contract.rs]: hanging onto the syscall:: prefix may help make this more readable

File: programs/storage_api/src/storage_contract.rs [hunk 2]
```diff
@@ -447,24 +470,48 @@ fn store_validation_result(
...
+            update_credits(credits, current.epoch);
+            credits.current_epoch += total_validations - recorded_validations;
         }
         _ => return Err(InstructionError::InvalidAccountData),
     }
     storage_account.account.set_state(&storage_contract)
 }
 
-fn count_valid_proofs(proofs: &[ProofStatus]) -> u64 {
+fn update_credits(credits: &mut Credits, current_epoch: u64) {
```
Comment (rob-solana) [inline programs/storage_api/src/storage_contract.rs]: this function looks like it would tally credits, but it's really just epoch-shifting?

Comment (rob-solana) [reply programs/storage_api/src/storage_contract.rs]: you didn't want to put this method on Credits?

File: programs/storage_api/src/storage_contract.rs [hunk 3]
```diff
@@ -9,14 +9,26 @@ use solana_sdk::hash::Hash;
...
+// Todo Tune this for actual use cases when PoRep is feature complete
 pub const STORAGE_ACCOUNT_SPACE: u64 = 1024 * 8;
 pub const MAX_PROOFS_PER_SEGMENT: usize = 80;
 
+#[derive(Default, Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct Credits {
+    // current epoch
+    epoch: u64,
+    // credits earned in the last epoch
+    pub last_epoch: u64,
```
Comment (rob-solana) [inline programs/storage_api/src/storage_contract.rs]: I can't find where last_epoch is used....

Comment (sagar-solana) [reply programs/storage_api/src/storage_contract.rs]: Unused, I don't know what to do with it right now.

File: programs/storage_api/src/storage_contract.rs (hunk 2 above)
Comment (sagar-solana) [reply programs/storage_api/src/storage_contract.rs]: You're right, better suited to live in Credits, update_epoch probably?

File: programs/storage_api/src/storage_contract.rs (hunk 3 above)
Comment (rob-solana) [reply programs/storage_api/src/storage_contract.rs]: delete it, then.

Comment (sagar-solana) [reply programs/storage_api/src/storage_contract.rs]: I thought I'd need it for something but after the implementation it ended up unused, I'll remove it.

**Comment by sagar-solana** [comment]
@rob-solana more comments :) ?

**Comment by sagar-solana** [comment]
I have not run tests yet, but most likely, I think the nodes will run of of lamports before they start getting payouts xD

**Comment by rob-solana** [comment]
maybe we make a loan ;)

**Comment by sagar-solana** [comment]
Yep, that or really its nbd they can just wait until the little work they were able to perform gets rewarded. It'll scale up eventually.

**Pull request merged**